File: src/types/engine.ts
// frontend/src/types/engine.ts
// Verbatim definitions verified against Python dataclasses in engine/contracts.py

export type TaskType =
  | 'single_image_vqa'
  | 'single_image_caption'
  | 'single_image_grounding'
  | 'temporal_change_detection'
  | 'temporal_change_description'
  | 'temporal_change_vqa'
  | 'cross_modal_optical_sar'
  | 'croma_classification';

export type InputType =
  | 'single_optical'
  | 'single_multispectral'
  | 'single_sar'
  | 'temporal_optical'
  | 'temporal_sar'
  | 'optical_sar_pair';

export type JobStatus =
  | 'queued'
  | 'running'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type Modality = 'optical' | 'sar';
export type Role = 'before' | 'after';
export type CoordinateType = 'pixel' | 'geo';

export interface BoundingBox {
  label: string; // e.g. 'water_agreement', 'changed_region', 'built_area'
  coordinates: [number, number, number, number]; // [xmin, ymin, xmax, ymax]
  coordinate_type: CoordinateType; // added by backend normalization
  confidence: number | null; // NULLABLE - preserved as null
  source: string; // 'optical' | 'sar' | 'cross_modal' | 'croma_classifier' | 'model'
}

export interface ChangeMask {
  width: number;
  height: number;
  mask_url: string | null; // backend converts mask_path to URL
  threshold_used: number | null;
  changed_pixel_count: number;
  changed_fraction: number;
}

export interface EvidenceBundle {
  textual_evidence: string | null;
  bounding_boxes: BoundingBox[];
  visualizations: string[]; // list of fully-qualified URLs
  change_statistics: Record<string, unknown> | null;
  change_mask: ChangeMask | null;
  metadata: Record<string, unknown>; // may contain fallback_triggered, fallback_reason
}

export interface EngineError {
  code: string; // 'PLANNING_FAILED' | 'INVALID_WORKFLOW' | 'MODEL_EXECUTION_FAILED' | etc.
  message: string;
}

export interface TraceStep {
  step: number;
  tool: string;
  task: string;
  status: string;
  parameters: Record<string, unknown>;
  duration_ms: number;
  result_summary: unknown;
}

export interface SpecialistResult {
  status: string;
  model_name: string;
  task: TaskType;
  answer: string | null;
  confidence: number | null; // nullable — NEVER replace with 0 or 1
  evidence: EvidenceBundle;
  metadata: Record<string, unknown>;
  execution_time: number; // seconds
  error: string | null;
}

export interface EngineResult {
  request_id: string;
  status: 'success' | 'failed';
  query: string;
  task: TaskType | null;
  answer: string | null;
  confidence: number | null; // nullable — NEVER replace with any default
  specialist_results: SpecialistResult[];
  evidence: EvidenceBundle[];
  execution_trace: TraceStep[];
  errors: EngineError[];
}

// API-level contracts
export interface AssetUploadResponse {
  asset_id: string;
  filename: string;
  size_bytes: number;
  format: string;
  width: number | null;
  height: number | null;
  bands: number | null;
  crs: string | null;
  resolution: number | null;
  bbox: [number, number, number, number] | null;
  preview_url?: string;
}

export interface JobSubmitRequest {
  query: string;
  assets: Array<{
    asset_id: string;
    modality: Modality;
    role?: Role;
    acquisition_time?: string;
  }>;
}

export interface JobResponse {
  job_id: string;
  status: JobStatus;
  created_at: string;
  updated_at: string;
  result: EngineResult | null;
}

export interface ApiError {
  code: string;
  message: string;
  details: unknown;
}

export interface CapabilitiesResponse {
  tasks: TaskType[];
  input_types: InputType[];
  supported_formats: string[];
  max_upload_bytes: number;
  models: {
    mock: string[];
    real: string[];
  };
}

export interface HealthResponse {
  status: string;
  version: string;
  engine_mode: string;
  timestamp: string;
}

File: backend/app/schemas/engine.py
"""Response schemas mirroring engine/contracts.py, with backend-side adaptations.

Rules (see SRS section 6/17):
- confidence fields are Optional[float] and must NEVER be coerced to 0/0.5/default.
- filesystem paths (mask_path, visualizations) are rewritten to API URLs before
  they ever reach these schemas.
- coordinate_type is added by the backend serializer, not present on the engine object.
"""
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class BoundingBoxResponse(BaseModel):
    label: str
    coordinates: list[float] = Field(description="[xmin, ymin, xmax, ymax] pixel or geo, see coordinate_type")
    coordinate_type: Literal["pixel", "geo"] = Field(description="Added by backend serializer")
    confidence: Optional[float] = Field(default=None, description="Nullable, never coerced")
    source: str = Field(description="optical | sar | cross_modal | croma_classifier")


class ChangeMaskResponse(BaseModel):
    width: int
    height: int
    mask_url: Optional[str] = Field(default=None, description="Rewritten from mask_path")
    threshold_used: Optional[float] = None
    changed_pixel_count: int
    changed_fraction: float = Field(description="0.0 to 1.0")


class EvidenceBundleResponse(BaseModel):
    textual_evidence: Optional[str] = None
    bounding_boxes: list[BoundingBoxResponse] = Field(default_factory=list)
    visualizations: list[str] = Field(default_factory=list, description="Rewritten to API URLs")
    change_statistics: Optional[dict[str, Any]] = None
    change_mask: Optional[ChangeMaskResponse] = None
    metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Passed through unchanged; may contain fallback_triggered / fallback_reason",
    )


class EngineErrorResponse(BaseModel):
    code: str
    message: str


class TraceStepResponse(BaseModel):
    step: int
    tool: str
    task: str
    status: str
    parameters: dict[str, Any] = Field(default_factory=dict)
    duration_ms: int
    result_summary: Any = Field(description="Redacted if it contains a filesystem path")


class SpecialistResultResponse(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    status: str
    model_name: str
    task: str
    answer: Optional[str] = None
    confidence: Optional[float] = Field(default=None, description="NULLABLE — never coerce")
    evidence: EvidenceBundleResponse
    metadata: dict[str, Any] = Field(default_factory=dict)
    execution_time: float
    error: Optional[str] = None


class EngineResultResponse(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    request_id: str
    status: str
    query: str
    task: Optional[str] = None
    answer: Optional[str] = None
    confidence: Optional[float] = Field(default=None, description="NULLABLE — never coerce")
    specialist_results: list[SpecialistResultResponse] = Field(default_factory=list)
    evidence: list[EvidenceBundleResponse] = Field(default_factory=list)
    execution_trace: list[TraceStepResponse] = Field(default_factory=list)
    errors: list[EngineErrorResponse] = Field(default_factory=list)
